### ISSUE TITLE
Hyperparameter Optimization of RL Models in FreqAI

### ISSUE THREAD

**Issue by chrootlogin** [freqtrade/freqtrade#10458]
## Describe your environment

  * Operating system: MacOS 14.4.1
  * Python Version: 3.12.4
  * CCXT version: 4.3.61
  * Freqtrade Version: 2024.7-dev-2560090a6
  
## Your question

I am currently using FreqAI with Reinforcement Learning (RL) and would like to optimize the hyperparameters of my RL model. While I have found that this can be done through parameter configuration, I am interested in using Optuna or a similar tool to find the optimal parameters based on my RL strategy. To do this, I need to initialize the RL model in the exact same way FreqAI does, but I am struggling with this.

Here’s what I have discovered so far:

- The RL model appears to be initialized in the _fit(self, data_dictionary: Dict[str, Any], dk: FreqaiDataKitchen, **kwargs)_ method of ReinforcementLearner.py, where the hyperparameters are also passed.
- The training occurs in the _step_ method of the actual RL model.

I would like to replicate this in a script to optimize the hyperparameters. For this, I need the data generated by FreqAI for a specific period, as it would be during backtesting/live or dry-run. However, the process of loading the data is fragmented and unclear to me as an outsider.

Here’s what I have figured out so far, but is not helpful, because it gives me zero features:
```python
from freqtrade.configuration import Configuration, TimeRange
from freqtrade.data.dataprovider import DataProvider
from freqtrade.freqai.data_kitchen import FreqaiDataKitchen
from freqtrade.resolvers import StrategyResolver, ExchangeResolver

config = Configuration.from_files(["****REDACTED****/user_data/config.json"])
config['timeframe'] = '5m'
config['timerange'] = '20240101-20240201'

exchange = ExchangeResolver.load_exchange(config)
strategy = StrategyResolver.load_strategy(config)
strategy.ft_bot_start()

strategy.dp = DataProvider(config, exchange, ["BTC/USDT"])
strategy.kitchen = FreqaiDataKitchen(config, False, "BTC/USDT")

freqai = strategy.freqai

timerange = TimeRange.parse_timerange(config['timerange'])

freqai.dd.load_all_pair_histories(timerange, strategy.kitchen)
freqai.dk = strategy.kitchen

corr_dataframes, base_dataframes = freqai.dd.get_base_and_corr_dataframes(
    timerange, "BTC/USDT", strategy.kitchen
)

unfiltered_dataframe = strategy.kitchen.use_strategy_to_populate_indicators(
    strategy, corr_dataframes, base_dataframes, "BTC/USDT"
)
```

To summarize, my question is: 

How can I initialize an RL model (e.g., the PPO model) so that it has the same data as generated by my FreqTrade strategy, allowing for easy manual tests or hyperparameter optimization?

Thank you for your assistance!

**Comment by xmatthias** [comment]
While i've assigned this issue to @robcaulk as the freqAI expert - i doubt that this is something he'll provide help with - so unless he'll be able to provide help in the next few days, i'll be closing this.

It's essentially a "extract logic from freqAI into an external script". 
While i don't mind that per se (it's open source) - that doesn't mean we'll have to spend our time in assisting users in doing so.
Quite some time has been spent on FreqAI, having it tested and everything working. 
By extracting partial logic and providing that as part of an issue answer, we'd essentially take (somewhat) responsibility to provide that correctly to replicate how freqAI does it - though without actually using the freqAI logics.
While you may say "oh i'll verify myself" - future readers of such an issue may not, or ignore that and expect an ever-valid snippet.

**Comment by robcaulk** [comment]
Most people use Optuna inside the Freqaimodel directly, for optimizing model hyperparamters. 

As Matthias says, extracting logic outside of Freqtrade isn't within the scope of how we intend FT to be used, and is therefore more difficult to support.

In this particular case, I see no reason not to use the typical approach of just calling Optuna inside your Freqaimodel file.

**Comment by chrootlogin** [comment]
Firstly, I want to thank @robcaulk and the team for their input. I understand now that my approach was misguided. My idea of loading the data separately was due to my overall understanding of the FreqAI architecture but finding the code structure somewhat complex. This complexity arises from the fact that certain parameters come from the config, others from the strategy, and yet others from the model configuration. But that's okay.

I would appreciate it if one of the developers could provide a small hint on where to best incorporate Optuna for optimization. Specifically, I am looking for a good entry point that ensures the data is sufficiently loaded and corresponds to what is used in backtesting, dry run, or live mode. Currently, I am considering whether it is appropriate to extend the extract_data_and_train_model(...) function, or if it would be better to incorporate it within the train(...) or fit(...) functions. A tip here would save me a lot of time, as I have already spent several hours looking through the code.

Ultimately, the idea would be to first train the model with the training data, then make predictions on the test data, calculate the reward, adjust the optimization parameters, retrain, test again, and finally select the model with the best reward.

I don't need the final code, just a list of functions with an explanation of where it would be best to address the problem so that I can implement it accordingly.

Thank you for your help!

**Comment by ekoshv** [comment]
Hi, I need to know too? But in total, changing from old to this perfect optimizer Optuna/TPE would be a great difference in the results.

**Comment by robcaulk** [comment]
@ekoshv I’ve provided some working optuna freqaimodel files to sponsors, but the gist is that it depends what you plan to optimize. It may be as simple as calling it inside “fit()” but it could require more complex usage if you need to optimize more parameters like train_period_days or test_size etc.

Feel free to use my files or work on building your own. Mine can be found here:

https://github.com/sponsors/robcaulk

**Comment by georgejosephada** [comment]
> [@ekoshv](https://github.com/ekoshv) I’ve provided some working optuna freqaimodel files to sponsors, but the gist is that it depends what you plan to optimize. It may be as simple as calling it inside “fit()” but it could require more complex usage if you need to optimize more parameters like train_period_days or test_size etc.
> 
> Feel free to use my files or work on building your own. Mine can be found here:
> 
> https://github.com/sponsors/robcaulk

where exactly are the files you are referring to please?

**Comment by robcaulk** [comment]
@cibic89 there are public files in the FreqAI discord, and there are private files reserved for sponsors. Please feel free to access whichever fits your system best.
